tests/engine/training/test_graph_trainer.py

```diff
@@ -0,0 +1,236 @@
+    train_with_schema: Callable,
+    spy_on_all_components: Callable,
+):
+
+    input_file = tmp_path / "input_file.txt"
+    input_file.write_text("3")
+
+    train_schema = GraphSchema(
+        {
+            "read_file": SchemaNode(
+                needs={},
+                uses=FileReader,
+                fn="read",
+                constructor_name="create",
+                config={"file_path": str(input_file)},
+                is_input=True,
+            ),
+            "train": SchemaNode(
+                needs={},
+                uses=PersistableTestComponent,
+                fn="train",
+                constructor_name="create",
+                config={"test_value": "4"},
+                is_target=True,
+            ),
+            "process": SchemaNode(
+                needs={"resource": "train"},
+                uses=PersistableTestComponent,
+                fn="run_inference",
+                constructor_name="load",
+                config={"wrap_output_in_cacheable": True},
+            ),
+            "add": SchemaNode(
+                needs={"i1": "read_file", "i2": "process"},
+                uses=AddInputs,
+                fn="add",
+                constructor_name="create",
+                config={},
+            ),
+            "assert_node": SchemaNode(
+                needs={"i": "add"},
+                uses=AssertComponent,
+                fn="run_assert",
+                constructor_name="create",
+                config={"value_to_assert": 7},
+                is_target=True,
+            ),
+        }
+    )
+
+    # The first train should call all the components
+    mocks = spy_on_all_components(train_schema)
+    train_with_schema(train_schema, temp_cache)
+    assert node_call_counts(mocks) == {
+        "read_file": 1,
+        "train": 1,
+        "process": 1,
+        "add": 1,
+        "assert_node": 1,
+    }
+
+    # Nothing has changed so this time so no components will run
+    # (just input nodes during fingerprint run).
+    mocks = spy_on_all_components(train_schema)
+    train_with_schema(train_schema, temp_cache)
+    assert node_call_counts(mocks) == {
+        "read_file": 1,  # Inputs nodes are always called during the fingerprint run.
+        "train": 0,
+        "process": 0,
+        "add": 0,
+        "assert_node": 0,
+    }
+
+    # As we changed the config of "add", all its descendants will run.
+    train_schema.nodes["add"].config["something"] = "new"
+    mocks = spy_on_all_components(train_schema)
+    train_with_schema(train_schema, temp_cache)
+    assert node_call_counts(mocks) == {
+        "read_file": 1,  # Inputs nodes are always called during the fingerprint run.
+        "train": 0,
+        "process": 0,
+        "add": 1,
+        "assert_node": 1,
+    }
+
+
+def node_call_counts(mocks: Dict[Text, Mock]) -> Dict[Text, int]:
+    return {node_name: mocks[node_name].call_count for node_name, mock in mocks.items()}
+
+
+@pytest.fixture
```

Comment:
i'd move all of these helpers up to the top of the module, what do you think?

Reply:
I find it more readable if you first read the tests, and then delve into the fixtures when/if appropriate. But don't feel super strongly.

Reply:
Yeah, makes sense sense, too. Happy with either way 👍🏻